docs/architecture/adr-082-data-companion-api.md

```diff
@@ -0,0 +1,425 @@
+// companion are most likely incorrectly configured.
+message UnexpectedFieldError {
+    // The name of the field whose value was expected to be empty, but was not.
+    string field_name = 1;
+}
+
+// ExpectedFieldError is returned by the server when Tendermint sent it a
+// message containing an empty field value for a field it was expecting to be
+// populated. If this occurs, it indicates that either Tendermint or the data
+// companion are most likely incorrectly configured.
+message ExpectedFieldError {
+    // The name of the field whose value was expected to be populated, but was
+    // empty.
+    string field_name = 1;
+}
+```
+
+### Request Buffering
+
+In order to ensure reliable delivery, while also catering for intermittent
+faults, we should facilitate buffering of data destined for the companion
+service. This data would need to be stored on disk to ensure that Tendermint
+could attempt to resubmit all unsubmitted data upon Tendermint startup.
+
+At present, this data is already stored on disk, and so practically we would
+need to implement some form of background pruning mechanism to remove the data
+we know has been shared with the companion service.
+
+In order to not allow disk storage to grow unbounded, _and_ ensure that we do
+not lose any critical data, we need to impose some kind of limit on the number
+of requests we will buffer before crashing the Tendermint node. In case there is
+manual intervention needed to restore the companion service, Tendermint should
+attempt to flush this buffer upon startup before continuing normal operations.
+
+### Configuration
+
+The following configuration file update is proposed to support the data
+companion API.
+
+```toml
+[data_companion]
+# By default, the data companion service interaction is disabled. It is
+# recommended that this only be enabled on full nodes and not validators so as
+# to minimize the possibility of network instability.
+enabled = false
+
+# Address at which the gRPC companion service server is hosted. It is
+# recommended that this companion service be co-located at least within the same
+# data center as the Tendermint node to reduce the risk of network latencies
+# interfering in node operation.
+addr = "http://localhost:26659"
+
+# Use the experimental gzip compression call option when submitting data to the
+# server. See https://pkg.go.dev/google.golang.org/grpc#UseCompressor
+experimental_use_gzip = false
+
+# Controls the BlockCommitted gRPC call.
+[data_companion.block_committed]
+# Enable the BlockCommitted gRPC call. Only relevant if the data companion is
+# enabled.
+enabled = true
+# Additional fields to publish in each BlockCommittedRequest sent to the
+# companion. Available options:
+# - "finalize_block_response": Also publish the FinalizeBlock response related
+#                              to the block in the BlockCommittedRequest.
+additionally_publish = ["finalize_block_response"]
```

Comment:
Why not a single list of data to be published, instead of `data_companion.block_committed` and `additionally_publish`?

Reply:
You're right, that's probably easier to deal with from a UX perspective. Instead of enabling different RPC calls (an implementation detail), rather let the user specify which kinds of data they want and let the companion client figure out which corresponding RPC calls and data structures to send.

I'll change the interface here.